.devcontainer/Dockerfile

```diff
@@ -0,0 +1,23 @@
+# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.245.2/containers/cpp/.devcontainer/base.Dockerfile
+
+# [Choice] Debian / Ubuntu version (use Debian 11, Ubuntu 18.04/22.04 on local arm64/Apple Silicon): debian-11, debian-10, ubuntu-22.04, ubuntu-20.04, ubuntu-18.04
+ARG VARIANT="bullseye"
+FROM mcr.microsoft.com/vscode/devcontainers/cpp:0-${VARIANT}
+
+# [Optional] Install CMake version different from what base image has already installed. 
+# CMake reinstall choices: none, 3.21.5, 3.22.2, or versions from https://cmake.org/download/
+ARG REINSTALL_CMAKE_VERSION_FROM_SOURCE="none"
+
+# Optionally install the cmake for vcpkg
+COPY ./reinstall-cmake.sh /tmp/
+RUN if [ "${REINSTALL_CMAKE_VERSION_FROM_SOURCE}" != "none" ]; then \
```

Comment:
If I understand correctly below, it seems a different setting of REINSTALL_CMAKE_VERSION_FROM_SOURCE overrides the one defined in this file, which makes this code relevant. I think it would be clearer if this file had the right setting right from the start, instead of having a bogus setting which is overriden in a separate file.

Reply:
> If I understand correctly, this code doesn't do anything if REINSTALL_CMAKE_VERSION_FROM_SOURCE is "none", which it is (a couple of lines above), so why do you need this code at all, or the "reinstall-cmake.sh" file?

> it seems a different setting of REINSTALL_CMAKE_VERSION_FROM_SOURCE overrides the one defined in this file, which makes this code relevant. I think it would be clearer if this file had the right setting right from the start, instead of having a bogus setting which is overriden in a separate file.

The setting in `ARG` is a default if none is specified, it is part of the base Dockerfile template

The `devcontainer.json` allows users to configure `ARG` parameters to override things like:
- The base OS image tag (ubuntu/debian + version)
- Any other `ARG` parameters in the scripts

This is for two reasons:
- It allows Docker to more intelligently cache layers/changes and prevent spurious rebuilds
- It allows users to override parameters in the devcontainer config rather than needing to modify the Dockerfile